usersession/userd/launcher.go

```diff
@@ -170,6 +185,216 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	// When running in a test environment dirs.SnapDesktopFilesDir is under /tmp/ and
+	// cannot pass these checks. Unless there's a way to subvert dirs.SnapDesktopFilesDir
+	// this hack should be harmless. Maybe there's a better way to fake it?
+	if !strings.HasPrefix(dirs.SnapDesktopFilesDir, "/tmp/") {
+
+		last := "/"
+		for _, val := range strings.Split(desktopFile, "/") {
+			checkPath := filepath.Join(last, val) // val is "" for root ('/')
+			fileStat, err := os.Stat(checkPath)
+			if err != nil || !(fileStat.Mode().IsDir() || fileStat.Mode().IsRegular()) || (fileStat.Mode().Perm()&0022) != 0 || fileStat.Sys().(*syscall.Stat_t).Uid != 0 {
+				return fmt.Errorf("cannot verify path %q", checkPath)
+			}
+			last = checkPath
+		}
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (string, error) {
+	var launch string
+
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return launch, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	in_desktop_section := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
+			in_desktop_section = true
+		} else if strings.HasPrefix(line, "[Desktop Action ") {
+			// maybe later we'll add support here
+			in_desktop_section = false
+		} else if strings.HasPrefix(line, "[") {
+			in_desktop_section = false
+		} else if in_desktop_section && strings.HasPrefix(line, "Exec=") {
+			launch = strings.TrimPrefix(line, "Exec=")
+			break
+		}
+	}
+
+	expectedPrefix := fmt.Sprintf("env BAMF_DESKTOP_FILE_HINT=%s /snap/bin/", desktopFile)
+	if !strings.HasPrefix(launch, expectedPrefix) {
+		return "", fmt.Errorf("Desktop file %q has an unsupported 'Exec' value: %q", desktopFile, launch)
+	}
+
+	return launch, nil
+}
+
+// Parse the Exec command by stripping any exec variables.
+// Passing exec variables (eg, %foo) between confined snaps is unsupported. Currently,
+// we do not have support for passing them in the D-Bus API but there are security
+// implications that must be thought through regarding the influence of the launching
+// snap over the launcher wrt exec variables. For now we simply filter them out.
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+func parseExecCommand(exec_command string) ([]string, error) {
+	args, err := shlex.Split(exec_command)
+	if err != nil {
+		return []string{}, err
+	}
+
+	i := 0
+	for {
+		// We want to keep literal '%' (expressed as '%%') but filter our exec variables
+		// like '%foo'
+		if strings.HasPrefix(args[i], "%%") {
+			args[i] = strings.TrimPrefix(args[i], "%")
+			i++
+		} else if strings.HasPrefix(args[i], "%") {
+			args = append(args[:i], args[i+1:]...)
```

Comment:
This looks like the correct thing to do for %f, %F, %u and %U.  Those aren't the only field codes available though:

https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables

I don't see the other field codes in use on any of the desktop files on my system, so not supporting them seems fine.  I think you should error out if you see any of them though.

Reply:
Alan made a change for this, marking as resolved.